packages/react-dom/src/events/DOMEventListenerSystem.js

```diff
@@ -0,0 +1,393 @@
+  nativeEvent: AnyNativeEvent,
+): void {
+  const eventPriority = getEventPriorityForType(eventType);
+  switch (eventPriority) {
+    case DiscreteEvent: {
+      flushDiscreteUpdatesIfNeeded(nativeEvent.timeStamp);
+      discreteUpdates(() =>
+        dispatchEventListenersByPhase(
+          captureTargetListenerInstances,
+          bubbleTargetListenerInstances,
+          rootCaptureListenerInstances,
+          rootBubbleListenerInstances,
+          nativeEvent,
+        ),
+      );
+      break;
+    }
+    case UserBlockingEvent: {
+      runWithPriority(UserBlockingPriority, () =>
+        dispatchEventListenersByPhase(
+          captureTargetListenerInstances,
+          bubbleTargetListenerInstances,
+          rootCaptureListenerInstances,
+          rootBubbleListenerInstances,
+          nativeEvent,
+        ),
+      );
+      break;
+    }
+    case ContinuousEvent: {
+      dispatchEventListenersByPhase(
+        captureTargetListenerInstances,
+        bubbleTargetListenerInstances,
+        rootCaptureListenerInstances,
+        rootBubbleListenerInstances,
+        nativeEvent,
+      );
+      break;
+    }
+  }
+}
+
+function getRootListenerInstances(
+  type: string,
+  passive: boolean,
+  capture: boolean,
+): Set<ReactListenerInstance<ReactDOMListener>> {
+  const rootEventListenerInstances = capture
+    ? rootCaptureEventListenerInstances
+    : rootBubbleEventListenerInstances;
+  let listenerInstances = rootEventListenerInstances.get(type);
+
+  if (listenerInstances === undefined) {
+    listenerInstances = {
+      active: new Set(),
+      passive: new Set(),
+    };
+    rootEventListenerInstances.set(type, listenerInstances);
+  }
+  return passive ? listenerInstances.passive : listenerInstances.active;
+}
+
+export function dispatchEventForListenerEventSystem(
+  eventType: string,
+  targetFiber: null | Fiber,
+  nativeEvent: AnyNativeEvent,
+  eventSystemFlags: EventSystemFlags,
+): void {
+  const isPassiveEvent = (eventSystemFlags & IS_PASSIVE) !== 0;
+  const isPassiveSupported = (eventSystemFlags & PASSIVE_NOT_SUPPORTED) === 0;
+  const isPassive = isPassiveEvent || !isPassiveSupported;
+  // Get target event listeners in their propagation order (non root events)
+  const {
+    capture: captureTargetListenerInstances,
+    bubble: bubbleTargetListenerInstances,
+  } = getTargetListenerInstances(eventType, targetFiber, isPassive);
+  const rootCaptureListenerInstances = arrayFrom(
+    getRootListenerInstances(eventType, isPassive, true),
+  );
+  const rootBubbleListenerInstances = arrayFrom(
+    getRootListenerInstances(eventType, isPassive, false),
+  );
+  // Fast-path to a no-op if we have no listener instances to process
+  if (
+    captureTargetListenerInstances.length === 0 &&
+    bubbleTargetListenerInstances.length === 0 &&
+    rootCaptureListenerInstances.length === 0 &&
+    rootBubbleListenerInstances.length === 0
+  ) {
```

Comment:
(We can also reverse the logic and put the dispatch inside the if statement to avoid the early return)